### PR TITLE
Updating Solid/Constitutive/GPfem/Umat

### PR DESCRIPTION
**Pull request by josep-m-carbonell**
1. new ConstitutiveModelsApplication (must be set ON for compilation of the SolidMechanics)
2. new UmatApplication (for geotechnical models)
3. update and reorganization of the SolidMechanicsApplication
4. update of the Pfem features 
5. submodule for the tetgen
6. needed compatibility with the GiDInteface for the SolidMechanicsApp and PfemApp
7. small changes in depending applications (must be checked for the responsible developers)

### CODE REVIEW

File: kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp
Comment (loumalouomega) [inline kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp]: Actually is better to use std::pow(val, 2)

File: applications/ConstitutiveModelsApplication/custom_models/constitutive_model_data.hpp
Comment (loumalouomega) [inline applications/ConstitutiveModelsApplication/custom_models/constitutive_model_data.hpp:170]: I recommend to use nullptr https://stackoverflow.com/questions/20509734/null-vs-nullptr-why-was-it-replaced

Can be replaced using clang tidy https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-nullptr.html

https://github.com/KratosMultiphysics/Kratos/wiki/How-to-use-Clang-Tidy-to-automatically-correct-code

File: applications/ConstitutiveModelsApplication/custom_models/elasticity_models/borja_model.cpp
Comment (loumalouomega) [inline applications/ConstitutiveModelsApplication/custom_models/elasticity_models/borja_model.cpp:69]: It is a good practice to use ++i instead of i++ https://stackoverflow.com/questions/24853/what-is-the-difference-between-i-and-i

File: applications/ConstitutiveModelsApplication/custom_models/elasticity_models/hyperelastic_model.cpp
Comment (loumalouomega) [inline applications/ConstitutiveModelsApplication/custom_models/elasticity_models/hyperelastic_model.cpp:193]: noalias?

File: applications/SolidMechanicsApplication/custom_elements/beam_elements/beam_element.cpp
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_elements/beam_elements/beam_element.cpp:115]: do you add always 3 displacements even in 2d?

File: applications/SolidMechanicsApplication/custom_elements/beam_elements/beam_element.cpp
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_elements/beam_elements/beam_element.cpp:1531]: Should this at least return a zero matrix?

**Comment by josep-m-carbonell** [comment]
@loumalouomega. Stop you comments please and focus in your affairs.

File: applications/SolidMechanicsApplication/custom_elements/beam_elements/beam_element.cpp
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_elements/beam_elements/beam_element.cpp:1544]: Should this at least return an zero value?

File: applications/SolidMechanicsApplication/custom_processes/add_dofs_process.h
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_processes/add_dofs_process.h:295]: The variables_utils can replace this process?

File: applications/SolidMechanicsApplication/custom_processes/assign_scalar_field_to_conditions_process.h
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_processes/assign_scalar_field_to_conditions_process.h:56]: can this process be replaced by the one in the core?

File: applications/SolidMechanicsApplication/custom_processes/assign_scalar_field_to_nodes_process.h
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_processes/assign_scalar_field_to_nodes_process.h:57]: idem

File: applications/SolidMechanicsApplication/custom_processes/assign_scalar_variable_to_conditions_process.h
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_processes/assign_scalar_variable_to_conditions_process.h:46]: idem

File: applications/SolidMechanicsApplication/custom_processes/assign_scalar_variable_to_nodes_process.h
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_processes/assign_scalar_variable_to_nodes_process.h:46]: idem

File: applications/SolidMechanicsApplication/custom_processes/assign_vector_field_to_conditions_process.h
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_processes/assign_vector_field_to_conditions_process.h:63]: idem

File: applications/SolidMechanicsApplication/custom_processes/assign_vector_variable_to_conditions_process.h
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_processes/assign_vector_variable_to_conditions_process.h:46]: idem

File: applications/SolidMechanicsApplication/custom_processes/fix_scalar_dof_process.h
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_processes/fix_scalar_dof_process.h:102]: can variables_utils replace this process?

File: applications/SolidMechanicsApplication/custom_processes/free_scalar_dof_process.h
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/custom_processes/free_scalar_dof_process.h:45]: can variables_utils replace this process?

File: applications/SolidMechanicsApplication/python_scripts/assign_modulus_and_direction_to_conditions_process.py
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/python_scripts/assign_modulus_and_direction_to_conditions_process.py:32]: can be replaced by the one in the core?

File: applications/SolidMechanicsApplication/python_scripts/assign_modulus_and_direction_to_nodes_process.py
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/python_scripts/assign_modulus_and_direction_to_nodes_process.py:27]: can be replaced by the one in the core?

File: applications/SolidMechanicsApplication/python_scripts/assign_scalar_to_nodes_process.py
Comment (loumalouomega) [inline applications/SolidMechanicsApplication/python_scripts/assign_scalar_to_nodes_process.py:35]: can be replaced by the one in the core?

**Comment by loumalouomega** [comment]
I added some commentaries, but certainly there is a lot of code to review

**Comment by loumalouomega** [comment]
I was trying to help

**Comment by josep-m-carbonell** [comment]
@loumalouomega. Thank you very much. You have done more than expected. Taking into account that nobody asked you as a reviewer. Even more, after doing parallel developmemts and copying code into another application in order to not contribute in SolidMechanics. Yourself can guess that you comments and opinions here are not appropiate.

File: .gitmodules
Comment (roigcarlo) [inline .gitmodules:3]: Not necessarily wrong but just some days ago we had to remove tetgen from the repo (see #906) for licencing issues.

To my understanding it now works as a external library like metis, trilinos etc... I honestly don't know if adding it as a submodule even if the code is hosted in your own repo ( which I understand it hold the original licence of tetgen ) is compatible with BSD.

**Review by lluis-mv — APPROVED** [review]
it is what we have been working on the fork and to me it is OK

**Review by roigcarlo — COMMENTED** [review]
Having a first look at the PR I have some concerns. I will let it compile to check everything is ok.

As a personal note, I understand that adding applications will generate huge PR's by definition but I would prefer to keep it as small as possible ( maybe 4 different PR, 1 per app?). It is far easier for everyone to look at the changes.

Also, reviews are open to **everyone**.

File: applications/UmatApplication/CMakeLists.txt
Comment (roigcarlo) [inline applications/UmatApplication/CMakeLists.txt:40]: UMAT seems to be using `KRATOS_API(UMAT_APPLICATION)` macro but `UMAT_APPLICATION` in not defined here. This will cause it will never export in windows. Don't know if its intended

**Comment by lorenzogracia** [comment]
I'm going to check the pull request in an external branch in order to verify if Dam and Poromechanics can be affected in some way.

File: applications/ConstitutiveModelsApplication/tests/generalTests.py
Comment (roigcarlo) [inline applications/ConstitutiveModelsApplication/tests/generalTests.py:11]: This test seems to do nothing. No need to include it just because its need to have one, If you intend to add tests latter it is ok.

File: applications/ConstitutiveModelsApplication/tests/test_ConstitutiveModelsApplication.py
Comment (roigcarlo) [inline applications/ConstitutiveModelsApplication/tests/test_ConstitutiveModelsApplication.py:12]: Same as above

File: .gitmodules
Comment (josep-m-carbonell) [reply .gitmodules:3]: Tetgen was not removed from the PfemApplication. Here the change is to replace it by a submodule. Therefore it is not in the repo anymore.

File: applications/UmatApplication/CMakeLists.txt
Comment (josep-m-carbonell) [reply applications/UmatApplication/CMakeLists.txt]: You explined to me that there were problems with the fortran compilation in windows. If they are solve I will thank you if add the mentioned correction.

Comment (josep-m-carbonell) [reply applications/UmatApplication/CMakeLists.txt:40]: You explained to me that there was some problems in fortran compilation for  windows. If they are solved I will thank you if you add the proper corrections.

Comment (lluis-mv) [reply applications/UmatApplication/CMakeLists.txt:40]: Hi, it is sort of intended. In this application we are using Fortran code, we thought it will be easier to compile in Linux and that's why we created this tiny application. Of course, this is almost a draft of the feature (be able to incorporate constitutive laws written in the Umat format, typical of comercial FEM softwares) that works but some more work is required to make it relevant.

File: applications/ConstitutiveModelsApplication/tests/generalTests.py
Comment (josep-m-carbonell) [reply applications/ConstitutiveModelsApplication/tests/generalTests.py:11]: We are buiding the tests, they are not included there.

File: applications/UmatApplication/CMakeLists.txt
Comment (roigcarlo) [reply applications/UmatApplication/CMakeLists.txt:40]: Well is not a problem per-se. It works with intel we just don't have the licence so we cannot use it, apart from that its just fine. 

In my opinion, it doesn't hurt to add the define as you done the hard part in the code, but I understand what you mean.

File: applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp
Comment (AFranci) [inline applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp:440]: Why do you want to modify this file? Are these the same changes that crashed the code some days ago?

File: applications/PfemApplication/custom_bounding/spatial_bounding_box.hpp
Comment (lluis-mv) [inline applications/PfemApplication/custom_bounding/spatial_bounding_box.hpp]: Hi, I think these four lines (from unsigned int start = 0 to start=1) may be removed since the variable start is not used and creates a compiler warning. KR

File: applications/PfemFluidDynamicsApplication/custom_processes/model_start_end_meshing_for_fluids_process.hpp
Comment (AFranci) [inline applications/PfemFluidDynamicsApplication/custom_processes/model_start_end_meshing_for_fluids_process.hpp:473]: Did you test if the pfem fluid still works after these modifications? If yes, for me it is ok

File: applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp
Comment (josep-m-carbonell) [reply applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp:440]: I hope the merge is done correctly. The changes were done to allow to print walls in GiD output. It was interferring in this process and this correction was needed.

**Review by AFranci — COMMENTED** [review]
Apart from the two highlighted concerns, the rest of the changes on the Pfem Fluid are fine

File: applications/PfemFluidDynamicsApplication/custom_processes/model_start_end_meshing_for_fluids_process.hpp
Comment (josep-m-carbonell) [reply applications/PfemFluidDynamicsApplication/custom_processes/model_start_end_meshing_for_fluids_process.hpp:473]: There are no tests in the application. I have tested some cases but I can not ensure all cases.

File: applications/PfemApplication/custom_bounding/spatial_bounding_box.hpp
Comment (josep-m-carbonell) [reply applications/PfemApplication/custom_bounding/spatial_bounding_box.hpp]: It is possible mistake. I have not detected it. Please can you add the correction to the branch?

**Comment by josep-m-carbonell** [comment]
Thank you @AFranci,  I will apreciate if you can compile it and check some of the classical cases.

File: applications/PfemApplication/custom_bounding/spatial_bounding_box.hpp
Comment (lluis-mv) [reply applications/PfemApplication/custom_bounding/spatial_bounding_box.hpp]: I think I have done it ;-P

File: applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp
Comment (AFranci) [reply applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp:440]: This function is for reducing adaptively the time step, I do not see how it can interfere with the GiD output. Anyway, there are changes of the function aim that I do not understand. Please, let the previous version of the function and, if required, we will modify in a new specific pull request.

Comment (josep-m-carbonell) [reply applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp:440]: Please try to comment the new lines and test the behaviour. Then you will understand why they are needed.

Comment (AFranci) [reply applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp:440]: we will discuss it the next week in the office, now I cannot test everything...

**Comment by pooyan-dadvand** [comment]
@loumalouomega Thanks for your review. Actually I agree with many of them. 

@josep-m-carbonell would you please break this pull request in some smaller ones ? The pull request is too large to be reviewed. (Is even difficult to see if the changes affect core or other applications)

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
I would just block it to avoid accidental merge.... so @josep-m-carbonell can divide it in smaller PRs.

File: applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp
Comment (josep-m-carbonell) [reply applications/PfemFluidDynamicsApplication/custom_processes/adaptive_time_interval_process.hpp:440]: To keep the same behaviour of the process wall elements must not be considered in the loop. This new elements are there for writing purposes. The original implementation is not aware of this matter and without this change the process fails.

File: kratos/includes/variables.h
Comment (roigcarlo) [inline kratos/includes/variables.h]: I strongly suggest giving this var another name as we have plenty of non-kratos variables and enums named `DIMENSION` in other apps.

**Comment by lorenzogracia** [comment]
Hi @josep-m-carbonell , I have created a branch in which I have merged your pull request with my branch (it is updated with the master), everything goes fine except a compilation error in the scheme ExplicitCentralDifferenesScheme where appears an error in pragma omp. I attach a snapshot. 

imagen

Can u check?

Regards

File: kratos/includes/variables.h
Comment (josep-m-carbonell) [reply kratos/includes/variables.h]: Well this is a  Kratos Variable used for other purposes. If there is any conflict in the future it can be changed.

**Comment by josep-m-carbonell** [comment]
Hi, @lorenzogracia, I know about that, but it does not happen to me, I think it is something related to the compiler in your ubuntu distribution, because the line seems correct. Maybe @roigcarlo can tell us a solution for that.

I will set mTime as private, not firstprivate. After the merge of the changes try to compile it again please.

**Comment by josep-m-carbonell** [comment]
@pooyan-dadvand , changes in the core included in this pull request are added separately in another pull request #963. So you will be able to check them. Other changes concern only to applications.

File: kratos/includes/variables.h
Comment (roigcarlo) [reply kratos/includes/variables.h]: The problem is that if we left it this way it does conflict, just the fact that it exists is a problem. For example, in DEM application there are multiple enums that define DIMENSION as a integer ( this one translates as a KratosVariable<int>). And probably it is not a easy to detect problem when/if it happens. 

Specifically what I amb afraid of is that some kind of partial shadowing could happend depending on which files are included or not. For example consider this code:

```C++
#include "variables.h"
#include "myapp.h" #but you don't include the file with the enum {DIMENSION = 3}

for(int i = 0; i < DIMENSION; i++) {
    doSomething();
}
```

This code probably compiles, and maybe only loop once, maybe it crashes, maybe loops 4 times and we end up geting a memleak.

Another solution would be to move it to your application or to pfem base / contact emchanichs if it needs to be inherited by anyone else. This way is isolated from other people, then the problem is only while coupling with other applications, which is perfectly fine.

Comment (josep-m-carbonell) [reply kratos/includes/variables.h]: Ok I see, the problem is critical and I have new questions:
1) This happens because these enums write the variables in CAPITALS which coincide with Kratos Variables standard ?  If the enum name is "Dimension" then no problem occurs?
(if it is like this, would be convienient to not use capitals for enum variables ?)
2) If a Kratos Variable and a Global or Local flag have the same name (they are usually written in capitals), have we got the same problem?
3) I think it is planned (in the future) to define the variables in the applications, not in Kratos core, then variables could be repeated. However it will be also dangerous because we can find the same variable defined as a local variable inside some classes.
I think you must expose this problem to @KratosMultiphysics/technical-committee .

Right now I will change the name to SPACE_DIMENSION to avoid conflicts and I will add the correction to this branch.

Comment (roigcarlo) [reply kratos/includes/variables.h]: 1) Yes, that is the problem indeed. As you said it won't be a problem if DIMENSION enum or C++ variables would have been called "Dimension" without caps, but enums usually are written in caps to show that are meaningful values.
2) Yes, if you are aware of any flag where this happens pls tell me.
3) Again, you are right. While in general it should not be the case this may happen. Probably the problem with dimension is that is somehow a very "generic" word. I recall having similar issues with VELOCITY for example. We will for sure discuss this in the next @KratosMultiphysics/technical-committee meeting.

**Comment by josep-m-carbonell** [comment]
Pull request of the core #963 was reviewed, corrected, approved and merged. Once the code compiles for @lorenzogracia. Please @pooyan-dadvand  unblock it.

**Comment by lorenzogracia** [comment]
Hi @josep-m-carbonell, still not compiles... I don't know why. @roigcarlo any idea of that?

imagen

**Comment by roigcarlo** [comment]
Yep, let me take a closer look when I arrive home but the error states that
mTime is not defined

**Comment by roigcarlo** [comment]
@lorenzogracia The problem here is that mTime is a struct, wich can only be private under certain circumstances. In this case `TimeVariables` needs to have a copy constructor, which is not explicitly declared (compiler will implicitly declare one) maybe the problem is somehow here.
See [This Link, section 25](https://software.intel.com/en-us/articles/32-openmp-traps-for-c-developers#IDAXSRDC) for more info.

Anyway, in the screen code there is no reason to consider mTime private as it is only read in the loop, so I would suggest to remove it if is giving problems. But the code seems fine...

**Comment by josep-m-carbonell** [comment]
Thank you @roigcarlo. Please @lorenzogracia , can you try compile without 'private(mTime)' and see if it works?  I have eliminated it.

**Comment by lorenzogracia** [comment]
Thank you @roigcarlo. Now it works.

**Review by lorenzogracia — APPROVED** [review]

**Comment by pooyan-dadvand** [comment]
Seems that it is working for @lorenzogracia and rest of the team. Only @maceligueta is has not confirmed. Meanwhile I unblock it to start making release branch.

**Review by pooyan-dadvand — APPROVED** [review]

**Pull request merged**
